Std/Data/BitVec/Basic.lean

```diff
@@ -0,0 +1,185 @@
+/-!
+We define bitvectors. We choose the `Fin` representation over others for its relative efficiency
+(Lean has special support for `Nat`), alignment with `UIntXY` types which are also represented
+with `Fin`, and the fact that bitwise operations on `Fin` are already defined. Some other possible
+representations are `List Bool`, `{ l : List Bool // l.length = w }`, `Fin w → Bool`.
+
+We also define many bitvector operations from the
+[`QF_BV` logic](https://smtlib.cs.uiowa.edu/logics-all.shtml#QF_BV).
+of SMT-LIBv2.
+
+TODO(Leo): match the interface to what SMT-LIB provides
+-/
+
+/--
+A bitvector of the specified width. This is represented as the underlying `Nat` number
+in both the runtime and the kernel, inheriting all the special support for `Nat`.
+-/
+structure BitVec (w : Nat) where
+  /-- Bitvector is represented as a `Fin`. -/
+  val : Fin (2^w)
+  deriving DecidableEq
+
+namespace BitVec
+
+theorem pow_two_gt_zero (w : Nat) : 2^w > 0 := by
+  apply Nat.pos_pow_of_pos; decide
+
+/-- The BitVector `i mod 2^n`. -/
+protected def ofNat (n : Nat) (i : Nat) : BitVec n :=
+  { val := Fin.ofNat' i (pow_two_gt_zero _) }
+
+/-- Given a bitvector `a`, return the underlying `Nat`. -/
+protected def toNat (a : BitVec n) : Nat :=
+  a.val.val
+
+/-- Return a bitvector `0` of size `n`.  -/
+protected def zero (n : Nat) : BitVec n :=
+  BitVec.ofNat n 0
+
+instance : Inhabited (BitVec n) where
+  default := BitVec.zero n
+
+instance : OfNat (BitVec n) i where
+  ofNat := BitVec.ofNat n i
+
+/--
+Notation for bit vector literals. `i#n` is a shorthand for `BitVec.ofNat n i`.
+TODO: is this a good notation? Do we need it? We can also write `(i : BitVec n)`
+-/
+scoped syntax:max num noWs "#" noWs num : term
+macro_rules | `($i:num#$n:num) => `(BitVec.ofNat $n $i)
+
+/-- Unexpander for bit vector literals. -/
+@[app_unexpander BitVec.ofNat] def unexpandBitVecOfNat : Lean.PrettyPrinter.Unexpander
+  | `($(_) $n:num $i:num) => `($i:num#$n:num)
+  | _                => throw ()
+
+instance : Repr (BitVec n) where
+  reprPrec a _ := repr a.val ++ "#" ++ repr n
+
+instance : ToString (BitVec n) where
+  toString a := toString (repr a)
+
+/-- Theorem for normalizing the bit vector literal representation. -/
+@[simp] theorem ofNat_eq_ofNat : @OfNat.ofNat (BitVec n) i _ = BitVec.ofNat n i := rfl
+
+/-- Addition for bit vectors. -/
+protected def add (x y : BitVec n) : BitVec n :=
+  { val := x.val + y.val }
+/-- Subtraction for bit vectors. -/
+protected def sub (x y : BitVec n) : BitVec n :=
+  { val := x.val - y.val }
+/-- Multiplication for bit vectors. -/
+protected def mul (x y : BitVec n) : BitVec n :=
+  { val := x.val * y.val }
+/-- Modulo for bit vectors. -/
+protected def mod (x y : BitVec n) : BitVec n :=
+  { val := ⟨x.val % y.val, Nat.lt_of_le_of_lt (Nat.mod_le _ _) x.val.isLt⟩ }
+/-- Division for bit vectors. -/
+protected def div (x y : BitVec n) : BitVec n :=
+  { val := ⟨x.val / y.val, Nat.lt_of_le_of_lt (Nat.div_le_self _ _) x.val.isLt⟩ }
```

Comment:
Can we reuse fin division here?

Reply:
Joe and Wojciech wrote their own version because it more efficient than the one in `Fin`. I will upstream this change to `Fin`, and use the `Fin` version here. Thanks!